### PR TITLE
BUG: have `_array_from_buffer_3118` correctly handle errors

### LINKED ISSUE #6741: ctypeslib.as_array aborts
This script fails:

```
import numpy as np
import ctypes

sh = (1, 1)
tmp = np.zeros(sh, np.uint8)
test = ctypes.cast(ctypes.c_void_p(tmp.__array_interface__['data'][0]), ctypes.POINTER(ctypes.c_ubyte))
np.ctypeslib.as_array(test, sh)
```

python 3.4.2-dbg
numpy==1.10.1
Ubuntu 14.10
1. Under release python this code runs without crashes, because of disabled assertions
2. Under debug release python asserts failed:
   `python3.4: ../Objects/abstract.c:2077: PyObject_Call: Assertion '(result != ((void *)0) && !PyErr_Occurred()) || (result == ((void *)0) && PyErr_Occurred())' failed.`

Can you please explain why?

P.S. may be it can help:
If `sh = (1, 1)` change to `sh = 239`:
1. Release python will raise `TypeError: shape must be a tuple`
2. Debug python will also aborts

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes gh-6741, closes gh-9348

If an object claims to support the PEP3118 buffer protocol, then any mistakes in the implementation should be raised as errors, rather than falling back on other approaches.

This has some behavior changes:

* np.array(mem_view_33d) now gives a more useful error message than "memoryview: unsupported format <B"
* np.array(ctypes_pointer) now errors rather than creating an object array containing a single pointer

Also adds support for 32-dimensional buffers, where previously the max was 31-dimensional.

### CODE REVIEW

File: numpy/core/tests/test_multiarray.py [hunk 1]
```diff
@@ -6515,6 +6518,31 @@ def test_out_of_order_fields(self):
...
 
+    def test_max_dims(self):
+        a = np.empty((1,) * 32)
+        self._check_roundtrip(a)
+
+    def test_error_too_many_dims(self):
+        # construct a memoryview with 33 dimensions
+        ct = ctypes.c_uint8
+        for i in range(33):
+            ct = 1 * ct
```
Comment (eric-wieser) [inline numpy/core/tests/test_multiarray.py]: Tests fail due to a reference cycle involving `ct`, which is somehow detected in a completely different test. Printing `gc.garbage` in the failing test gives:
```
# ndim = 18
A = numpy.core.tests.test_multiarray.c_ubyte_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1
B = numpy.core.tests.test_multiarray.c_ubyte_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1_Array_1

[<class 'A'>,
 (<class 'A'>,
  <class '_ctypes.Array'>,
  <class '_ctypes._CData'>,
  <class 'object'>),
 {'__dict__': <attribute '__dict__' of 'A' objects>,
  '__doc__': None,
  '__module__': 'numpy.core.tests.test_multiarray',
  '__weakref__': <attribute '__weakref__' of 'A' objects>,
  '_length_': 1,
  '_type_': <class 'B'>},
 <attribute '__weakref__' of 'A' objects>,
 <attribute '__dict__' of 'A' objects>]
```

This happens even if we remove the `np.array` call below, so the problem must be in either CPython, or parallelization in pytest

**Comment by charris** [comment]
Thanks Eric.

**Comment by eric-wieser** [comment]
I'm not 100% confident that this the behavior this patch fixes wasn't something that we rely upon. In particular, perhaps `__array_interface__` is supposed to take priority over the buffer protocol.

**Comment by charris** [comment]
Yeah, I figure we will find out, and if not, it probably doesn't matter. But if you come upon some info that suggests it does matter, please update. These changes in behavior are tricky, but sometimes worth exploring.

**Comment by eric-wieser** [comment]
Yep, this broke `np.ctypeslib.as_array(p, shape)` - #10968

**Pull request merged**